In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [5]:
df = pd.read_csv("./data/heart_2022_with_nans.csv")
df.head(10)

State     Sex GeneralHealth  PhysicalHealthDays  MentalHealthDays  \
0  Alabama  Female     Very good                 0.0               0.0   
1  Alabama  Female     Excellent                 0.0               0.0   
2  Alabama  Female     Very good                 2.0               3.0   
3  Alabama  Female     Excellent                 0.0               0.0   
4  Alabama  Female          Fair                 2.0               0.0   
5  Alabama    Male          Poor                 1.0               0.0   
6  Alabama  Female     Very good                 0.0               0.0   
7  Alabama  Female          Good                 0.0               0.0   
8  Alabama  Female          Good                 0.0               0.0   
9  Alabama  Female          Good                 1.0               0.0   

                                     LastCheckupTime PhysicalActivities  \
0  Within past year (anytime less than 12 months ...                 No   
1                                                NaN                 No   
2  Within past year (anytime less than 12 months ...                Yes   
3  Within past year (anytime less than 12 months ...                Yes   
4  Within past year (anytime less than 12 months ...                Yes   
5  Within past year (anytime less than 12 months ...                 No   
6  Within past year (anytime less than 12 months ...                Yes   
7  Within past year (anytime less than 12 months ...                 No   
8  Within past year (anytime less than 12 months ...                Yes   
9  Within past year (anytime less than 12 months ...                Yes   

   SleepHours RemovedTeeth HadHeartAttack  ... HeightInMeters  \
0         8.0          NaN             No  ...            NaN   
1         6.0          NaN             No  ...           1.60   
2         5.0          NaN             No  ...           1.57   
3         7.0          NaN             No  ...           1.65   
4         9.0          NaN             No  ...           1.57   
5         7.0          NaN            Yes  ...           1.80   
6         7.0          NaN             No  ...           1.65   
7         8.0          NaN             No  ...           1.63   
8         6.0          NaN             No  ...           1.70   
9         7.0          NaN             No  ...           1.68   

  WeightInKilograms    BMI AlcoholDrinkers HIVTesting FluVaxLast12  \
0               NaN    NaN              No         No          Yes   
1             68.04  26.57              No         No           No   
2             63.50  25.61              No         No           No   
3             63.50  23.30              No         No          Yes   
4             53.98  21.77             Yes         No           No   
5             84.82  26.08              No         No           No   
6             62.60  22.96             Yes         No           No   
7             73.48  27.81              No         No          Yes   
8               NaN    NaN              No        Yes           No   
9             81.65  29.05             Yes        NaN          Yes   

  PneumoVaxEver                                  TetanusLast10Tdap  \
0            No  Yes, received tetanus shot but not sure what type   
1            No  No, did not receive any tetanus shot in the pa...   
2            No                                                NaN   
3           Yes  No, did not receive any tetanus shot in the pa...   
4           Yes  No, did not receive any tetanus shot in the pa...   
5           Yes  No, did not receive any tetanus shot in the pa...   
6            No  No, did not receive any tetanus shot in the pa...   
7           Yes  Yes, received tetanus shot but not sure what type   
8            No  Yes, received tetanus shot but not sure what type   
9           Yes  No, did not receive any tetanus shot in the pa...   

  HighRiskLastYear CovidPos  
0               No       No  
1               No       No  
2               

In [6]:
df.info() #Hay variables tanto categóricas, como numéricas como booleanas.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445132 entries, 0 to 445131
Data columns (total 40 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   State                      445132 non-null  object 
 1   Sex                        445132 non-null  object 
 2   GeneralHealth              443934 non-null  object 
 3   PhysicalHealthDays         434205 non-null  float64
 4   MentalHealthDays           436065 non-null  float64
 5   LastCheckupTime            436824 non-null  object 
 6   PhysicalActivities         444039 non-null  object 
 7   SleepHours                 439679 non-null  float64
 8   RemovedTeeth               433772 non-null  object 
 9   HadHeartAttack             442067 non-null  object 
 10  HadAngina                  440727 non-null  object 
 11  HadStroke                  443575 non-null  object 
 12  HadAsthma                  443359 non-null  object 
 13  HadSkinCancer              44

In [7]:
df.shape

(445132, 40)

In [8]:
df.columns

Index(['State', 'Sex', 'GeneralHealth', 'PhysicalHealthDays',
       'MentalHealthDays', 'LastCheckupTime', 'PhysicalActivities',
       'SleepHours', 'RemovedTeeth', 'HadHeartAttack', 'HadAngina',
       'HadStroke', 'HadAsthma', 'HadSkinCancer', 'HadCOPD',
       'HadDepressiveDisorder', 'HadKidneyDisease', 'HadArthritis',
       'HadDiabetes', 'DeafOrHardOfHearing', 'BlindOrVisionDifficulty',
       'DifficultyConcentrating', 'DifficultyWalking',
       'DifficultyDressingBathing', 'DifficultyErrands', 'SmokerStatus',
       'ECigaretteUsage', 'ChestScan', 'RaceEthnicityCategory', 'AgeCategory',
       'HeightInMeters', 'WeightInKilograms', 'BMI', 'AlcoholDrinkers',
       'HIVTesting', 'FluVaxLast12', 'PneumoVaxEver', 'TetanusLast10Tdap',
       'HighRiskLastYear', 'CovidPos'],
      dtype='object')

In [9]:
print (df["Sex"].value_counts())
print (df["Sex"].value_counts(normalize=True)*100)

Sex
Female    235893
Male      209239
Name: count, dtype: int64
Sex
Female    52.993943
Male      47.006057
Name: proportion, dtype: float64


In [10]:
df["RaceEthnicityCategory"].value_counts()

RaceEthnicityCategory
White only, Non-Hispanic         320421
Hispanic                          42917
Black only, Non-Hispanic          35446
Other race only, Non-Hispanic     22713
Multiracial, Non-Hispanic          9578
Name: count, dtype: int64

|Columna|Descripción|Tipo|Categorías|
|:-----:|:-----:|:-----:|:-----:|
|State|Estado de residencia de la persona encuestada en USA|Categórica|Estados de USA|
|Sex|Género de la persona encuestada|Categórica|Male,Female|
|General Health|Estado general de la salud|Categórica|Poor,Fair,Good,Very good,Excellent|
|Physical Health Days|Días con mala salud física de los últimos 30 días|Numérica|Días(0-30)|
|Mental Health Days|Días con mala salud mental de los últimos 30 días|Numérica|Días(0-30)|
|Last Checkup Time|Tiempo desde el último chequeo|Categórica|<1 year, 1-2 years, >5 years, Never|
|Physical Activities|Si realiza actividad física o no|Booleana|Yes/No|
|Sleep Hours|Horas de sueño diario|Numérica|Horas|
|Removed Teeth|Cantidad de dientes quitados|Categórica|NaN,1-5,6+,All|
|Had Heart Attach|Si ha sufrido o no un ataque al corazón|Booleana|Yes/No|
|Had Angina|Si ha sufrido una angina de pecho|Booleana|Yes/No|
|Had Stroke|Si ha sufrido un derrame cerebral|Booleana|Yes/No|
|Had Asthma|Si tiene asma o no|Booleana|Yes/No|
|Had Skin Cancer|Si ha sufrido cáncer de piel|Booleana|Yes/No|
|Had COPD|Si tiene EPOC|Booleana|Yes/No|
|Had Depressiva Disorder|Si ha sido diagnosticado de trastorno depresivo|Booleana|Yes/No|
|Had Kidney Disease|Si tiene problemas renales|Booleana|Yes/No|
|Had Arthritis|Diagnóstico de artritis|Booleana|Yes/No|
|Had Diabetes|Diagnóstico de diabetes|Categórica|Yes, No, Borderline, During Pregnancy|
|Deaf Or Hard Of Hearing|Problemas auditivos|Booleana|Yes/No|
|Blind Or Vision Difficulty|Problemas visuales|Booleana|Yes/No|
|Difficulty Concentrating|Dificultad para concentrarse|Booleana|Yes/No|
|Difficulty Walking|Dificultad para andar|Booleana|Yes/No|
|Difficulty Dressing Bathing|Dificultad para vestirse o bañarse|Booleana|Yes/No|
|Difficulty Errands|Dificultad para realizar recados solo|Booleana|Yes/No|
|Smoker Status|Estado respecto al hábito de fumar|Categórica|Never, Former, Current|
|ECigarette Usage|Uso de cigarrillos electrónicos|Categórica|Never, Former, Current|
|Chest Scan|Se se ha realizado un escáner torácico|Booleana|Yes/No|
|Race Ethnicity Category|Grupo racial o étnico|Categórica|White, Black, Hispanic, etc.|
|Age Category|Edad de la persona encuestada|Numérica|Edad|
|Height In Meters|Altura de la persona encuestada|Numérica|Metros|
|Weight In Kilograms|Peso de la persona encuestada|Numérica|kg|
|BMI|Índice de Masa Corporal|Numérica|kg|
|Alcohol Drinkers|Consumo de alcohol|Booleana|Yes/No|
|HIV Testing|Si se ha realizado el test de VIH|Booleana|Yes/No|
|Flu Vax Last 12|Si se ha puesto la vacuna de la gripe en los últimos 12 meses|Booleana|Yes/No|
|Pneumo Vax Ever|Si le han puesto la vacuna neumocócica alguna vez|Booleana|Yes/No|
|Tetanus Last 10 Tdap|Si le han puesto la vacuna antitetánica (Tdap) en los últimos 10 años|Categórica|Yes, No, Don't know|
|High Risk Last Year|Conductas de alto riesgo durante el último año|Booleana|Yes/No|
|CovidPos|Resultado positivo en una prueba de COVID-19|Categórica|Yes, No, Don't know|



In [11]:
total_nulos = df.isna().sum().sum()
total_nulos

np.int64(902665)

In [12]:
print ((df.isna().sum().sort_values(ascending=False)))


TetanusLast10Tdap            82516
PneumoVaxEver                77040
HIVTesting                   66127
ChestScan                    56046
CovidPos                     50764
HighRiskLastYear             50623
BMI                          48806
FluVaxLast12                 47121
AlcoholDrinkers              46574
WeightInKilograms            42078
ECigaretteUsage              35660
SmokerStatus                 35462
HeightInMeters               28652
DifficultyErrands            25656
DifficultyConcentrating      24240
DifficultyWalking            24012
DifficultyDressingBathing    23915
BlindOrVisionDifficulty      21564
DeafOrHardOfHearing          20647
RaceEthnicityCategory        14057
RemovedTeeth                 11360
PhysicalHealthDays           10927
AgeCategory                   9079
MentalHealthDays              9067
LastCheckupTime               8308
SleepHours                    5453
HadAngina                     4405
HadSkinCancer                 3143
HadHeartAttack      

In [13]:
total_celdas = df.shape[0]*df.shape[1]
porcentaje_nulos = (total_nulos/total_celdas)*100
round (porcentaje_nulos,2)

np.float64(5.07)

In [14]:
print (df.duplicated().sum())

157


In [15]:
df.drop_duplicates(inplace=True)

# **Notas**

- En este primer análisis del dataset, podemos observar que hay un total de 40 columnas y 445.132 filas. 
- El total de valores nulos es de 889.788. La categoría que tiene más nulos es la de Tetanus Last 10Tdap.
- Encontramos un total de 157 valores duplicados, por lo que hacemos un drop.duplicates para eliminarlos.
- El porcentaje de respuesta entre hombres y mujeres estñá muy igualado (53% vs 47%)

In [16]:
df.fillna("UNK", inplace=True)

C:\Users\judit\AppData\Local\Temp\ipykernel_22436\600810073.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'UNK' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.fillna("UNK", inplace=True)


In [17]:
pd.set_option("display.max_columns", 100)
pd.set_option("display.max_colwidth", None) #Los valores que son largos y se ven con ... los pone completos
pd.set_option("display.max_rows", 100)

df.head(5)



State     Sex GeneralHealth PhysicalHealthDays MentalHealthDays  \
0  Alabama  Female     Very good                0.0              0.0   
1  Alabama  Female     Excellent                0.0              0.0   
2  Alabama  Female     Very good                2.0              3.0   
3  Alabama  Female     Excellent                0.0              0.0   
4  Alabama  Female          Fair                2.0              0.0   

                                      LastCheckupTime PhysicalActivities  \
0  Within past year (anytime less than 12 months ago)                 No   
1                                                 UNK                 No   
2  Within past year (anytime less than 12 months ago)                Yes   
3  Within past year (anytime less than 12 months ago)                Yes   
4  Within past year (anytime less than 12 months ago)                Yes   

  SleepHours RemovedTeeth HadHeartAttack HadAngina HadStroke HadAsthma  \
0        8.0          UNK             No        No        No        No   
1        6.0          UNK             No        No        No        No   
2        5.0          UNK             No        No        No        No   
3        7.0          UNK             No        No        No       Yes   
4        9.0          UNK             No        No        No        No   

  HadSkinCancer HadCOPD HadDepressiveDisorder HadKidneyDisease HadArthritis  \
0            No      No                    No               No           No   
1           Yes      No                    No               No           No   
2           Yes      No                    No               No           No   
3            No      No                    No               No          Yes   
4            No      No                    No               No           No   

  HadDiabetes DeafOrHardOfHearing BlindOrVisionDifficulty  \
0         Yes                  No                      No   
1          No                  No                      No   
2          No                  No                      No   
3          No                  No                      No   
4          No                  No                      No   

  DifficultyConcentrating DifficultyWalking DifficultyDressingBathing  \
0                      No                No                        No   
1                      No                No                        No   
2                      No                No                        No   
3                      No                No                        No   
4                      No                No                        No   

  DifficultyErrands                           SmokerStatus  \
0                No                           Never smoked   
1                No                           Never smoked   
2                No                           Never smoked   
3                No  Current smoker - now smokes some days   
4                No                           Never smoked   

                             ECigaretteUsage ChestScan  \
0                     Not at all (right now)        No   
1  Never used e-cigarettes in my entire life        No   
2  Never used e-cigarettes in my entire life        No   
3  Never used e-cigarettes in my entire life       Yes   
4  Never used e-cigarettes in my entire life       Yes   

      RaceEthnicityCategory      AgeCategory HeightInMeters WeightInKilograms  \
0  White only, Non-Hispanic  Age 80 or older            UNK               UNK   
1  White only, Non-Hispanic  Age 80 or older            1.6             68.04   
2  White only, Non-Hispanic     Age 55 to 59           1.57              63.5   
3  White only, Non-Hispanic              UNK           1.65              63.5   
4  White only, Non-Hispanic     Age 40 to 44           1.57             53.98   

     BMI AlcoholDrinkers HIVTesting FluVaxLast12 PneumoVaxEver  \
0    UNK              No         No          Yes            No   
1  26.57              No         No           No

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 444975 entries, 0 to 445131
Data columns (total 40 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   State                      444975 non-null  object
 1   Sex                        444975 non-null  object
 2   GeneralHealth              444975 non-null  object
 3   PhysicalHealthDays         444975 non-null  object
 4   MentalHealthDays           444975 non-null  object
 5   LastCheckupTime            444975 non-null  object
 6   PhysicalActivities         444975 non-null  object
 7   SleepHours                 444975 non-null  object
 8   RemovedTeeth               444975 non-null  object
 9   HadHeartAttack             444975 non-null  object
 10  HadAngina                  444975 non-null  object
 11  HadStroke                  444975 non-null  object
 12  HadAsthma                  444975 non-null  object
 13  HadSkinCancer              444975 non-null  objec

# **Notas**

- Hemos decidido reemplazar los valores NaN por "UNK". De esta forma al hacer un .info() ya nos muestra que no queda ningún nulo.
- Para que aparezca visualmente mejor en VSCode, hemos introducido el siguiente código:

  * pd.set_option("display.max_columns", 100)

  * pd.set_option("display.max_colwidth", None) #Los valores que son largos y se ven con ... los pone completos

  * pd.set_option("display.max_rows", 100)